### PR TITLE
feat: serverless producer for Lambda/Edge environments

### LINKED ISSUE #89: feat: serverless/edge-compatible producer - enqueue without persistent connections
## Problem

Serverless functions (AWS Lambda, Cloudflare Workers, Vercel Edge) and edge runtimes have short lifespans and connection limits. Creating a full Queue instance with a persistent Valkey connection for each invocation is wasteful and may not even be possible in edge runtimes that lack TCP socket support.

## Evidence of demand

- **BullMQ discussion** — 18 votes (top discussion), specifically about Lambda/serverless
- Serverless is the dominant deployment model for many teams
- Edge computing is growing rapidly (Cloudflare Workers, Vercel Edge, Deno Deploy)

## Current state

`Queue` requires a persistent `GlideClient` or `GlideClusterClient` connection. No lightweight producer mode, no HTTP-based enqueue, no connection pooling strategy for short-lived environments.

### Option A: Lightweight producer mode
```ts
import { LightQueue } from 'glide-mq';

// Opens connection, enqueues, closes - optimized for single-use
const queue = new LightQueue('emails', { connectionString: 'valkey://...' });
await queue.add('welcome', { to: 'user@example.com' });
await queue.close();
```

### Option B: HTTP proxy (pairs with #83)
```ts
// No Valkey connection needed - calls HTTP endpoint
const queue = new HttpQueue('emails', { endpoint: 'https://mq.myapp.com' });
await queue.add('welcome', { to: 'user@example.com' });
```

### Option C: Connection reuse in serverless
```ts
// Reuse connection across warm Lambda invocations
let queue: Queue;
export const handler = async (event) => {
  if (!queue) queue = new Queue('emails', { connection });
  await queue.add('welcome', event.body);
};
```

## Implementation notes

- Option A: optimize connection lifecycle - fast connect, enqueue, disconnect
- Option B: depends on HTTP proxy (#83)
- Option C: documentation + best practices guide for serverless
- Consider: Valkey connection overhead (TLS handshake, AUTH, FUNCTION LOAD check)
- Edge runtimes may need a fetch-based adapter instead of TCP sockets

## Labels

`tier-3`

### PR DESCRIPTION
**Pull request by avifenesh**
## Summary

Implements lightweight Producer class for serverless/edge environments (AWS Lambda, Cloudflare Workers, Vercel Edge Functions) where persistent connections are expensive or impossible.

### Key Changes

- **Producer class** - Fire-and-forget job enqueueing without EventEmitter overhead
  - Returns string job IDs (not Job instances)
  - Supports both ephemeral and pooled clients
  - Reuses battle-tested addJob/dedup FCALL functions
  - API: `add()`, `addBulk()`, `close()`

- **ServerlessPool** - Connection reuse for warm invocations
  - Caches Producer instances by queue name + connection fingerprint
  - Module-level singleton for convenient use
  - API: `getProducer()`, `closeAll()`, `size`

- **Documentation** - Comprehensive serverless guide
  - docs/SERVERLESS.md with AWS Lambda, Cloudflare Workers, Vercel Edge examples
  - README.md updated with serverless section
  - CHANGELOG.md entries for v0.8.1

### Implementation Details

- No breaking changes - Producer is a new export
- Zero new Lua code - reuses existing Queue functions
- Batched cross-queue parent registrations (O(1) vs O(N) round trips)
- Consistent GlideMQError usage throughout
- Compression included in pool fingerprint to prevent cache collisions

### Test Coverage

- 58 tests total (29 standalone + 29 cluster mode)
- Tests for Producer.add(), addBulk(), close()
- Tests for ServerlessPool caching and lifecycle
- Integration tests (Producer + Worker)
- Edge cases: dedup, compression, validation, errors

### Review Notes

- All review findings addressed (3 critical, 2 high, 8 medium)
- Deslop clean (no AI slop detected)
- TypeScript compilation clean
- All tests passing

## Test Plan

- [x] npm run build - TypeScript compilation
- [x] npm test - Full test suite passing
- [x] Tests in both standalone and cluster modes
- [x] Producer + Worker integration verified
- [x] Deduplication, compression, validation tested

## Related Issues

Closes #89

### CODE REVIEW

File: tests/broadcast.test.ts [hunk 1]
```diff
@@ -301,13 +301,15 @@ describeEachMode('Broadcast with dedup integration', (CONNECTION) => {
 
     // Publish with dedup ID - dedup is configured via JobOptions
     const id1 = await broadcast.publish(
+      'message',
       { event: 'deduped' },
       { deduplication: { id: 'unique-1', mode: 'simple', ttl: 5000 } },
     );
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply tests/broadcast.test.ts]: This review was based on an older version of the code. The broadcast-subject-filter changes that introduced publish(subject, data, opts) have been reverted. The current implementation uses the original publish(data, opts?) 2-arg API. Tests now correctly call broadcast.publish(data, opts?).

File: tests/broadcast.test.ts [hunk 2]
```diff
@@ -301,13 +301,15 @@ describeEachMode('Broadcast with dedup integration', (CONNECTION) => {
...
     expect(id1).not.toBeNull();
 
     // Duplicate - should be skipped
     const id2 = await broadcast.publish(
+      'message',
       { event: 'deduped' },
       { deduplication: { id: 'unique-1', mode: 'simple', ttl: 5000 } },
     );
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply tests/broadcast.test.ts]: Same as above - reviewed on an outdated version. broadcast-subject-filter has been reverted; current API is publish(data, opts?).

File: tests/broadcast.test.ts [hunk 3]
```diff
@@ -244,7 +244,7 @@ describeEachMode('Broadcast with scheduler integration', (CONNECTION) => {
     await Promise.all([worker1.waitUntilReady(), worker2.waitUntilReady()]);
 
     // Schedule a message with 2 second delay
-    await broadcast.publish({ event: 'scheduled' }, { delay: 2000 });
+    await broadcast.publish('message', { event: 'scheduled' }, { delay: 2000 });
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply tests/broadcast.test.ts]: Same - old review. API reverted to publish(data, opts?).

File: src/broadcast.ts [hunk 4]
```diff
@@ -56,12 +56,14 @@ export class Broadcast<D = any> extends EventEmitter {
...
+   * @param subject - Message subject/topic for routing (e.g., 'projects.1.issues.2').
+   *   BroadcastWorker `subjects` patterns match against this value.
    * @param data - Message data
    * @param opts - Job options (delay, priority, dedup, etc.)
    * @returns Message ID or null if skipped (e.g., due to dedup)
    */
-  async publish(data: D, opts?: JobOptions): Promise<string | null> {
-    const job = await this.queue.add('message', data, opts);
+  async publish(subject: string, data: D, opts?: JobOptions): Promise<string | null> {
+    const job = await this.queue.add(subject, data, opts);
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply src/broadcast.ts]: Correct observation - the broadcast-subject-filter merge introduced a breaking API change. This has been reverted: publish(subject, data, opts) is now back to publish(data, opts?) in this PR.

File: src/serverless-pool.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+   * Returns a cached instance if one exists with matching connection parameters.
+   *
+   * Note: Injected clients and custom serializers bypass caching to prevent
+   * collisions where different instances could map to the same cache key.
+   */
+  getProducer<D = any>(name: string, opts: ProducerOptions): Producer<D> {
+    // Bypass caching for injected clients or custom serializers to prevent cache collisions
+    if (opts.client || opts.serializer) {
+      return new Producer(name, opts);
+    }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply src/serverless-pool.ts]: Same design as previous replies: serializer-bypass producers are intentionally caller-managed one-off instances.

File: src/serverless-pool.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+   * Get or create a Producer for the given queue name and options.
+   * Returns a cached instance if one exists with matching connection parameters.
+   *
+   * Note: Injected clients and custom serializers bypass caching to prevent
+   * collisions where different instances could map to the same cache key.
+   */
+  getProducer<D = any>(name: string, opts: ProducerOptions): Producer<D> {
+    // Bypass caching for injected clients or custom serializers to prevent cache collisions
+    if (opts.client || opts.serializer) {
+      return new Producer(name, opts);
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply src/serverless-pool.ts]: Same design decision - serializer-bypass producers are one-off instances managed by the caller.
